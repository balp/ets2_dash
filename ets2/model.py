import json
from dataclasses import dataclass
from typing import Optional, Dict, List

from ets2.config import TruckConfig, truckconfig_from_dict, TrailerConfig, trailer_config_from_dict
from ets2.jobs import JobConfig, jobconfig_from_dict
from ets2.telematic import Telematic, telematic_from_dict
from ets2.tracks import Tracks


@dataclass
class Info:
    paused: bool


def info_from_dict(data: Dict) -> Info:
    return Info(paused=data['paused'])


@dataclass
class Version:
    major: int
    minor: int


def version_from_dict(data: Dict) -> Version:
    return Version(data['major'], data['minor'])


@dataclass
class Game:
    id: str
    name: str
    raw_version: int
    version: Version


def game_from_dict(data: Dict) -> Game:
    return Game(id=data['id'],
                name=data['name'],
                raw_version=data['raw_version'],
                version=version_from_dict(data['version']))


class Model:
    def __init__(self):
        self._observers = []
        self.telematic: Optional[Telematic] = None
        self.job: Optional[JobConfig] = None
        self.info: Optional[Info] = None
        self.game: Optional[Game] = None
        self.truck_config: Optional[TruckConfig] = None
        self.trailer_config: List[Optional[TrailerConfig]] = [None for _ in range(10)]
        self.tracks: Tracks = Tracks()

    def register_observer(self, observer):
        self._observers.append(observer)

    def _data_changed(self, change):
        for observer in self._observers:
            observer.notify(self, change)

    def set_telematic_data(self, data):
        self.telematic = telematic_from_dict(data)
        self.tracks.add_telematic(self.telematic)
        self._data_changed("telematic")

    def set_job_config(self, data):
        self.job = jobconfig_from_dict(data)
        self._data_changed("job")

    def set_info(self, data):
        self.info = info_from_dict(data)
        self._data_changed("info")

    def set_game(self, data):
        self.game = game_from_dict(data)
        self._data_changed("game")

    def set_truck_config(self, data):
        self.truck_config = truckconfig_from_dict(data)
        self._data_changed("truck_config")

    def set_trailer_config(self, data, index):
        self.trailer_config[index] = trailer_config_from_dict(data)
        self._data_changed("trailer_config")

    def get_game_pause(self) -> Optional[bool]:
        if self.info:
            return self.info.paused
        return None

    def get_game_name(self) -> Optional[str]:
        if self.game:
            return self.game.name
        return None

    def get_game_id(self) -> Optional[str]:
        if self.game:
            return self.game.id
        return None

    def get_time_left(self) -> Optional[int]:
        if self.telematic and self.job:
            game_time = self.telematic.common.game_time
            delivery_time: int = self.job.delivery_time
            if game_time and delivery_time and delivery_time != 0xFFFFFFFF:
                return delivery_time - game_time
        return None

    def get_time_to_rest(self) -> int:
        if self.telematic:
            return self.telematic.common.rest_stop
        return 0

    def get_time_destination(self) -> int:
        if self.telematic:
            eta = self.telematic.truck.navigation_time
            return int(eta) // 60
        return 0

    def get_time_destination_with_rest(self) -> int:
        if self.telematic:

            drive_time = self.get_time_destination()
            next_rest = self.get_time_to_rest()
            total_time = drive_time

            if drive_time > next_rest:
                total_time += 9*60
                total_time += ((drive_time - next_rest) // (11 * 60)) * 9 * 60
            return int(total_time)
        return 0

    def get_speed_kmh(self) -> float:
        if self.telematic:
            return self.telematic.truck.speed * 3.6
        return 0.0

    def get_cruise_control_kmh(self) -> float:
        if self.telematic:
            return self.telematic.truck.cruise_control * 3.6
        return 0.0

    def get_speed_limit_kmh(self) -> float:
        if self.telematic:
            return self.telematic.truck.navigation_speed_limit * 3.6
        return 0.0

    def get_speed_mph(self) -> float:
        if self.telematic:
            return self.telematic.truck.speed * 2.2369363
        return 0.0

    def get_cruise_control_mph(self) -> float:
        if self.telematic:
            return self.telematic.truck.cruise_control * 2.2369363
        return 0.0

    def get_speed_limit_mph(self) -> float:
        if self.telematic:
            return self.telematic.truck.navigation_speed_limit * 2.2369363
        return 0.0

    def get_fuel_left(self) -> float:
        if self.telematic:
            return self.telematic.truck.fuel_amount
        return 0.0

    def get_fuel_range(self) -> float:
        if self.telematic:
            return self.telematic.truck.fuel_range
        return 0.0

    def get_fuel_consumtion(self) -> float:
        if self.telematic:
            return self.telematic.truck.fuel_consumption_average
        return 0.0

    def get_wear_cabin(self) -> float:
        if self.telematic:
            return self.telematic.truck.wear_cabin
        return 0.0

    def get_wear_chassis(self) -> float:
        if self.telematic:
            return self.telematic.truck.wear_chassis
        return 0.0

    def get_wear_engine(self) -> float:
        if self.telematic:
            return self.telematic.truck.wear_engine
        return 0.0

    def get_wear_transmission(self) -> float:
        if self.telematic:
            return self.telematic.truck.wear_transmission
        return 0.0

    def get_wear_wheels(self) -> float:
        if self.telematic:
            return self.telematic.truck.wear_wheels
        return 0.0

    def get_wear_trailer(self) -> float:
        if self.telematic:
            return self.telematic.trailer.wear_chassis
        return 0.0

    def get_light_high_beam(self) -> bool:
        if self.telematic:
            return self.telematic.truck.light_beam_high
        return False

    def get_light_low_beam(self) -> bool:
        if self.telematic:
            return self.telematic.truck.light_beam_low
        return False

    def get_light_beacon(self) -> bool:
        if self.telematic:
            return self.telematic.truck.light_beacon
        return False

    def get_light_l_blinker(self) -> bool:
        if self.telematic:
            return self.telematic.truck.light_lblinker
        return False

    def get_light_r_blinker(self) -> bool:
        if self.telematic:
            return self.telematic.truck.light_rblinker
        return False

    def get_l_blinker(self) -> bool:
        if self.telematic:
            return self.telematic.truck.lblinker
        return False

    def get_r_blinker(self) -> bool:
        if self.telematic:
            return self.telematic.truck.rblinker
        return False

    def get_light_parking(self) -> bool:
        if self.telematic:
            return self.telematic.truck.light_parking
        return False

    def get_light_reverse(self) -> bool:
        if self.telematic:
            return self.telematic.truck.light_reverse
        return False

    def get_light_aux_front(self) -> bool:
        if self.telematic:
            return self.telematic.truck.light_aux_front != 0
        return False

    def get_light_aux_roof(self) -> bool:
        if self.telematic:
            return self.telematic.truck.light_aux_roof != 0
        return False

    def get_light_breaking(self) -> bool:
        if self.telematic:
            return self.telematic.truck.light_brake
        return False

    def get_ad_blue_warning(self) -> bool:
        if self.telematic:
            return self.telematic.truck.adblue_warning
        return False

    def get_break_emergency(self) -> bool:
        if self.telematic:
            return self.telematic.truck.brake_air_pressure_emergency
        return False

    def get_break_warning(self) -> bool:
        if self.telematic:
            return self.telematic.truck.brake_air_pressure_warning
        return False

    def get_break_motor(self) -> bool:
        if self.telematic:
            return self.telematic.truck.brake_motor
        return False

    def get_break_parking(self) -> bool:
        if self.telematic:
            return self.telematic.truck.brake_parking
        return False

    def get_electric(self) -> bool:
        if self.telematic:
            return self.telematic.truck.electric_enabled
        return False

    def get_battery_warning(self) -> bool:
        if self.telematic:
            return self.telematic.truck.battery_voltage_warning
        return False

    def get_engine(self) -> bool:
        if self.telematic:
            return self.telematic.truck.engine_enabled
        return False

    def get_fuel_warning(self) -> bool:
        if self.telematic:
            return self.telematic.truck.fuel_warning
        return False

    def get_oil_warning(self) -> bool:
        if self.telematic:
            return self.telematic.truck.oil_pressure_warning
        return False

    def get_water_warning(self) -> bool:
        if self.telematic:
            return self.telematic.truck.water_temperature_warning
        return False

    def get_wipers(self) -> bool:
        if self.telematic:
            return self.telematic.truck.wipers
        return False

    def get_air_pressure(self) -> float:
        if self.telematic:
            return self.telematic.truck.brake_air_pressure
        return False

    def get_break_retarder(self) -> int:
        if self.telematic:
            return self.telematic.truck.brake_retarder
        return False

    def get_break_temperature(self) -> float:
        if self.telematic:
            return self.telematic.truck.brake_temperature
        return False


def add_json_to_model(model: Model, json_data: json, topic: str):
    if topic == "ets2/data":
        model.set_telematic_data(json_data)
    elif topic == "ets2/game":
        model.set_game(json_data)
    elif topic == "ets2/info":
        model.set_info(json_data)
    elif topic == "ets2/info/config/job":
        model.set_job_config(json_data)
    elif topic == "ets2/info/config/truck":
        model.set_truck_config(json_data)
    elif topic == "ets2/info/config/trailer.0":
        model.set_trailer_config(json_data, 0)
    elif topic == "ets2/info/config/trailer.1":
        model.set_trailer_config(json_data, 1)
    elif topic == "ets2/info/config/trailer.2":
        model.set_trailer_config(json_data, 2)
    elif topic == "ets2/info/config/trailer.3":
        model.set_trailer_config(json_data, 3)
    elif topic == "ets2/info/config/trailer.4":
        model.set_trailer_config(json_data, 4)
    elif topic == "ets2/info/config/trailer.5":
        model.set_trailer_config(json_data, 5)
    elif topic == "ets2/info/config/trailer.6":
        model.set_trailer_config(json_data, 6)
    elif topic == "ets2/info/config/trailer.7":
        model.set_trailer_config(json_data, 7)
    elif topic == "ets2/info/config/trailer.8":
        model.set_trailer_config(json_data, 8)
    elif topic == "ets2/info/config/trailer.9":
        model.set_trailer_config(json_data, 9)
