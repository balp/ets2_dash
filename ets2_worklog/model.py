import datetime
import json
from dataclasses import dataclass
from typing import Optional, Dict, List

import ets2.model


@dataclass
class Delivered:
    auto_load: bool
    auto_park: bool
    cargo_damage: float
    time: int
    distance: float
    xp: int
    revenue: int


def delivered_from_dict(data: Dict) -> Delivered:
    return Delivered(auto_load=data['auto.load.used'], auto_park=data['auto.park.used'],
                     cargo_damage=data['cargo.damage'], time=data['delivery.time'],
                     distance=data['distance.km'], xp=data['earned.xp'], revenue=data['revenue'])


@dataclass
class Cancelled:
    penalty: int


def cancelled_from_dict(data: Dict) -> Cancelled:
    return Cancelled(penalty=data['cancel.penalty'])


@dataclass
class Job:
    config: Optional[ets2.model.JobConfig]
    started: Optional[datetime.datetime]
    ended: Optional[datetime.datetime]
    delivered: Optional[Delivered]
    cancelled: Optional[Cancelled]
    track: ets2.model.Tracks


class WorkLog:
    """A log work jobs in ETS2/ATS"""

    def __init__(self, data: ets2.model.Model):
        self._model = data
        self._model.register_observer(self)
        self._jobs: List[Job] = []

    def __repr__(self):
        return str(self._jobs)

    def notify(self, model: ets2.model.Model, change: str):
        if self._jobs:
            if model.job != self._jobs[-1].config:
                self._jobs[-1].ended = self.time_in_game(model)
                job = self.job_from_model(model)
                self._jobs.append(job)
            else:
                if model.telematic:
                    self._jobs[-1].track.add_telematic(model.telematic)
        else:
            job = self.job_from_model(model)
            self._jobs.append(job)

    def job_from_model(self, model):
        started = self.time_in_game(model)
        job = Job(config=model.job,
                  started=started,
                  ended=None,
                  delivered=None,
                  cancelled=None,
                  track=ets2.model.Tracks())
        return job

    def time_in_game(self, model):
        if model.telematic:
            started = datetime.datetime.fromordinal(1) \
                      + datetime.timedelta(minutes=model.telematic.common.game_time)
        else:
            started = None
        return started

    def job_delivered(self, delivered: Delivered) -> None:
        if self._jobs[-1].config:
            self._jobs[-1].delivered = delivered

    def job_cancelled(self, cancelled: Cancelled) -> None:
        if self._jobs[-1].config:
            self._jobs[-1].cancelled = cancelled


def add_json_to_work_log(work_log: WorkLog, json_data: json, topic: str) -> None:
    if topic == "ets2/info/gameplay/job.cancelled":
        work_log.job_cancelled(cancelled_from_dict(json_data))
    elif topic == "ets2/info/gameplay/job.delivered":
        work_log.job_delivered(delivered_from_dict(json_data))
