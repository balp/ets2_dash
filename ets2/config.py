#  Copyright (c) 2020. Anders Arnholm <Anders@Arnholm.se>
#
#   Permission to use, copy, modify, and distribute this software for any
#   purpose with or without fee is hereby granted, provided that the above
#   copyright notice and this permission notice appear in all copies.
#
#   THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
#   WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
#   MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
#   ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
#   WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
#   ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
#   OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
from dataclasses import dataclass
from typing import Dict, Optional

from ets2.types import Vector, vector_from_dict


@dataclass
class TruckConfig:
    id: str
    name: str
    adblue_capacity: float
    adblue_warning_factor: float
    battery_voltage_warning: float
    brake_air_pressure_emergency: float
    brake_air_pressure_warning: float
    brand: str
    brand_id: str
    cabin_position: Vector
    differential_ratio: float
    forward_ratio: float
    fuel_capacity: float
    fuel_warning_factor: float
    gears_forward: int
    gears_reverse: int
    head_position: Vector
    hook_position: Vector
    license_plate: str
    license_plate_country: str
    license_plate_country_id: str
    oil_pressure_warning: float
    retarder_steps: int
    reverse_ratio: float
    rpm_limit: float
    water_temperature_warning: float
    wheel_position: Vector
    wheel_powered: bool
    wheel_radius: float
    wheel_simulated: bool
    wheel_steerable: bool
    wheels_count: bool


def truckconfig_from_dict(data: Dict) -> Optional[TruckConfig]:
    if len(data):
        return TruckConfig(
            id=data['id'],
            name=data['name'],
            adblue_capacity=data['adblue.capacity'],
            adblue_warning_factor=data['adblue.warning.factor'],
            battery_voltage_warning=data['battery.voltage.warning'],
            brake_air_pressure_emergency=data['brake.air.pressure.emergency'],
            brake_air_pressure_warning=data['brake.air.pressure.warning'],
            brand=data['brand'],
            brand_id=data['brand_id'],
            cabin_position=vector_from_dict(data['cabin.position']),
            differential_ratio=data['differential.ratio'],
            forward_ratio=data['forward.ratio'],
            fuel_capacity=data['fuel.capacity'],
            fuel_warning_factor=data['fuel.warning.factor'],
            gears_forward=data['gears.forward'],
            gears_reverse=data['gears.reverse'],
            head_position=vector_from_dict(data['head.position']),
            hook_position=vector_from_dict(data['hook.position']),
            license_plate=data['license.plate'],
            license_plate_country=data['license.plate.country'],
            license_plate_country_id=data['license.plate.country.id'],
            oil_pressure_warning=data['oil.pressure.warning'],
            retarder_steps=data['retarder.steps'],
            reverse_ratio=data['reverse.ratio'],
            rpm_limit=data['rpm.limit'],
            water_temperature_warning=data['water.temperature.warning'],
            wheel_position=vector_from_dict(data['wheel.position']),
            wheel_powered=data['wheel.powered'],
            wheel_radius=data['wheel.radius'],
            wheel_simulated=data['wheel.simulated'],
            wheel_steerable=data['wheel.steerable'],
            wheels_count=data['wheels.count'],
        )
    else:
        return None


@dataclass
class TrailerConfig:
    id: str
    body_type: Optional[str]
    chain_type: Optional[str]
    cargo_accessory_id: str
    hook_position: Vector
    license_plate: str
    license_plate_country: str
    license_plate_country_id: str
    wheel_position: Vector
    wheel_powered: bool
    wheel_radius: float
    wheel_simulated: bool
    wheel_steerable: bool
    wheels_count: int


def trailer_config_from_dict(data: Dict) -> Optional[TrailerConfig]:
    if len(data):
        return TrailerConfig(
            id=data['id'],
            body_type=(data['body.type'] if 'body.type' in data else None),
            chain_type=(data['chain.type'] if 'chain.type' in data else None),
            cargo_accessory_id=data['cargo.accessory.id'],
            hook_position=vector_from_dict(data['hook.position']),
            license_plate=data['license.plate'],
            license_plate_country=data['license.plate.country'],
            license_plate_country_id=data['license.plate.country.id'],
            wheel_position=vector_from_dict(data['wheel.position']),
            wheel_powered=data['wheel.powered'],
            wheel_radius=data['wheel.radius'],
            wheel_simulated=data['wheel.simulated'],
            wheel_steerable=data['wheel.steerable'],
            wheels_count=data['wheels.count'],
        )
    else:
        return None