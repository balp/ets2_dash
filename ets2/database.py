import sqlite3
from typing import List

from ets2.jobs import Job, Delivered, Cancelled, JobConfig
from ets2.types import Vector
from ets2.tracks import Tracks, TrackPoint


def _get_job_delivered(curr2, job, job_id):
    curr2.execute('select auto_load, auto_park, cargo_damage, time,'
                  ' distance, xp, revenue from job_delivered where id=?',
                  (job_id,))
    r = curr2.fetchone()
    if r is not None:
        print(f"Got delivered {r}")
        job.delivered = Delivered(auto_load=r[0],
                                  auto_park=r[1],
                                  cargo_damage=r[2],
                                  time=r[3],
                                  distance=r[4],
                                  xp=r[5],
                                  revenue=r[6])


def _get_job_config(curr2, job, job_id):
    curr2.execute('select cargo, cargo_id, cargo_mass,'
                  ' delivery_time, destination_city, destination_city_id,'
                  ' destination_company, destination_company_id,'
                  ' income, source_city, source_city_id,'
                  ' source_company, source_company_id from job_config where id=?',
                  (job_id,))
    r = curr2.fetchone()
    if r is not None:
        print(f"Got config {r}")
        job.config = JobConfig(cargo=r[0],
                               cargo_id=r[1],
                               cargo_mass=r[2],
                               delivery_time=r[3],
                               destination_city=r[4],
                               destination_city_id=r[5],
                               destination_company=r[6],
                               destination_company_id=r[7],
                               income=r[8],
                               source_city=r[9],
                               source_city_id=r[10],
                               source_company=r[11],
                               source_company_id=r[12])


def _get_job_cancelled(curr2, job, job_id):
    curr2.execute('select penalty from job_cancelled where id=?',
                  (job_id,))
    r = curr2.fetchone()
    if r is not None:
        print(f"Got config {r}")
        job.cancelled = Cancelled(penalty=r[0])


def _get_job_track(curr2, job, job_id):
    curr2.execute('select last_time from tracks where id=?',
                  (job_id,))
    r = curr2.fetchone()
    if r is not None:
        print(f"Got config {r}")
        job.track.last_time = r[0]

    for count, time, x, y, z in curr2.execute('select count, time, x, y, z from track_point'
                                              ' where id=?'
                                              ' order by count', (job_id,)):
        job.track.points.append(TrackPoint(position=Vector(x=x, y=y, z=z), time=time))


class DataBase:
    def __init__(self, db_path, db_name):
        db_path.mkdir(parents=True, exist_ok=True)
        database = db_path / db_name
        print(f"DataBase.__init__(... {database.absolute()} ...)")
        self._conn = sqlite3.connect(database.absolute(), isolation_level=None)
        cursor = self._conn.cursor()
        cursor.executescript("""
            CREATE TABLE IF NOT EXISTS job
            (
                id      integer primary key,
                started integer,
                ended   integer
            );
            
            CREATE TABLE IF NOT EXISTS job_config
            (
                id                     integer primary key,
                cargo                  text,
                cargo_id               text,
                cargo_mass             real,
                delivery_time          integer,
                destination_city       text,
                destination_city_id    text,
                destination_company    text,
                destination_company_id text,
                income                 integer,
                source_city            text,
                source_city_id         text,
                source_company         text,
                source_company_id      text,
                FOREIGN KEY (id)
                    REFERENCES job (id)
            );
            
            CREATE TABLE IF NOT EXISTS job_delivered
            (
                id           integer primary key,
                auto_load    integer,
                auto_park    integer,
                cargo_damage real,
                time         integer,
                distance     real,
                xp           integer,
                revenue      integer,
                FOREIGN KEY (id)
                    REFERENCES job (id)
            );
            
            CREATE TABLE IF NOT EXISTS job_cancelled
            (
                id      integer primary key,
                penalty integer,
                FOREIGN KEY (id)
                    REFERENCES job (id)
            );
            
            CREATE TABLE IF NOT EXISTS tracks
            (
                id    integer primary key,
                last_time integer,
                FOREIGN KEY (id)
                    REFERENCES job (id)
            );
            
            CREATE TABLE IF NOT EXISTS track_point
            (
                id        integer,
                count     integer,
                time      integer,
                x real,
                y real,
                z real,
                PRIMARY KEY (id, count),
                FOREIGN KEY (id)
                    REFERENCES tracks (id)
            );
""")
        cursor.close()

    def save_job(self, job: Job) -> int:
        print(f"Save job {job.id}: {job.started}")
        if job is None:
            return -1
        cursor = self._conn.cursor()
        if job.id is not None:
            print(f"update job set started = {job.started}, ended = {job.ended} where id = {job.id}")
            cursor.execute("update job set started = ?, ended = ? where id = ?",
                           (job.started, job.ended, job.id))
            job_id = job.id
        else:
            print(f"insert into job (started, ended) values ({job.started},{job.ended})")
            cursor.execute("insert into job (started, ended) values (?,?)",
                           (job.started, job.ended))
            job.id = cursor.lastrowid
            job_id = job.id
        if job.config is not None:
            cursor.execute("replace into job_config (id, cargo, cargo_id, cargo_mass, delivery_time, "
                           "                         destination_city, destination_city_id, destination_company,"
                           "                         destination_company_id, income, source_city, source_city_id,"
                           "                         source_company, source_company_id)"
                           "   values (?,?,?,?,?,?,?,?,?,?,?,?,?,?)",
                           (job.id,
                            job.config.cargo,
                            job.config.cargo_id,
                            job.config.cargo_mass,
                            job.config.delivery_time,
                            job.config.destination_city,
                            job.config.destination_city_id,
                            job.config.destination_company,
                            job.config.destination_company_id,
                            job.config.income,
                            job.config.source_city,
                            job.config.source_city_id,
                            job.config.source_company,
                            job.config.source_company_id))
        if job.delivered is not None:
            cursor.execute("replace into job_delivered ("
                           "    id, auto_load, auto_park, cargo_damage,"
                           "    time, distance, xp, revenue)"
                           "   values (?,?,?,?,?,?,?,?)",
                           (job.id,
                            job.delivered.auto_load,
                            job.delivered.auto_park,
                            job.delivered.cargo_damage,
                            job.delivered.time,
                            job.delivered.distance,
                            job.delivered.xp,
                            job.delivered.revenue))
        if job.cancelled is not None:
            cursor.execute("replace into job_cancelled (id, penalty) values (?,?)",
                           (job.id, job.cancelled.penalty))
        cursor.execute("replace into tracks (id, last_time) values (?,?)", (job.id, job.track.last_time))
        for i, point in enumerate(job.track.points):
            cursor.execute("replace into track_point (id, count, time, x, y, z) values (?,?,?,?,?,?)",
                           (job.id, i, point.time, point.position.x, point.position.y, point.position.z))
        cursor.close()
        return job_id

    def get_jobs(self) -> List[Job]:
        jobs: List[Job] = []
        curr = self._conn.cursor()
        for job_id, started, ended in curr.execute('select id, started, ended from job order by id'):
            print(f"job_id '{job_id}' {type(job_id)}, started '{started}', ended '{ended}'")
            job = Job(config=None, started=started, ended=ended, delivered=None, cancelled=None,
                      track=Tracks(), id=job_id)
            curr2 = self._conn.cursor()
            _get_job_config(curr2, job, job_id)
            _get_job_delivered(curr2, job, job_id)
            _get_job_cancelled(curr2, job, job_id)
            _get_job_track(curr2, job, job_id)

            jobs.append(job)
            curr2.close()
        curr.close()
        return jobs
