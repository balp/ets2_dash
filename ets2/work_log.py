import datetime
import json
from pathlib import Path
from typing import List, Optional

import ets2.model
import ets2.tracks
from ets2.database import DataBase
from ets2.jobs import Job, Delivered, delivered_from_dict, Cancelled, cancelled_from_dict


def time_in_game(model: ets2.model.Model):
    if model.telematic:
        started = game_time_to_datetime(model.telematic.common.game_time)
    else:
        started = None
    return started


def game_time_to_datetime(time):
    return datetime.datetime.fromordinal(1) \
           + datetime.timedelta(minutes=time)


def job_from_model(model):
    game_time = None
    if model.telematic:
        game_time = model.telematic.common.game_time
    return Job(id=None,
               config=model.job,
               started=game_time,
               ended=None,
               delivered=None,
               cancelled=None,
               track=ets2.tracks.Tracks())


def _game_time_in_model(model: ets2.model.Model) -> Optional[int]:
    game_time = None
    if model.telematic:
        game_time = model.telematic.common.game_time
    return game_time


class WorkLog:
    """A log work jobs in ETS2/ATS"""

    def __init__(self, data: ets2.model.Model, database: Optional[DataBase]) -> None:
        self._model = data
        self._model.register_observer(self)
        if database is None:
            self._db = DataBase(Path.home() / '.local/share/ets2_work_log/',
                                'work_log.sqlite')
        else:
            self._db = database
        self.jobs: List[Job] = self._db.get_jobs()

    def __repr__(self):
        return str(self.jobs)

    def notify(self, model: ets2.model.Model, _: str):
        print(f"notify:({model.job}) {len(self.jobs)}:")
        if model is None:
            return
        if model.telematic is None:
            return
        if len(self.jobs) > 0:
            current_job = self.jobs[-1]
            if model.job != current_job.config:
                print(f"start new job")
                current_job.ended = _game_time_in_model(model)
                self._db.save_job(current_job)
                self._add_new_job(model)
            else:
                if model.telematic:
                    if current_job.track.add_telematic(model.telematic):
                        self._db.save_job(current_job)
        else:
            print(f"new first job")
            self._add_new_job(model)

    def _add_new_job(self, model):
        new_job = job_from_model(model)
        self.jobs.append(new_job)
        self._db.save_job(new_job)

    def job_delivered(self, delivered: Delivered) -> None:
        if self.jobs[-1].config:
            self.jobs[-1].delivered = delivered

    def job_cancelled(self, cancelled: Cancelled) -> None:
        if self.jobs[-1].config:
            self.jobs[-1].cancelled = cancelled


def add_json_to_work_log(work_log: WorkLog, json_data: json, topic: str) -> None:
    if topic == "ets2/info/gameplay/job.cancelled":
        work_log.job_cancelled(cancelled_from_dict(json_data))
    elif topic == "ets2/info/gameplay/job.delivered":
        work_log.job_delivered(delivered_from_dict(json_data))
