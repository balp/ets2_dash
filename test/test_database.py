from approvaltests import verify

from ets2.model import Model
from ets2.work_log import WorkLog
from test.utils import rerun_data_from_files, verify_work_log_as_json


def test_save_job():
    _, _, database = rerun_data_from_files(["data/ats_start_port_ang_coos_bay.mqtt.json.bz2"],
                                           'test_save_job')
    lines = []
    for line in database._conn.iterdump():
        lines.append(line)
    verify('\n'.join(lines))


def test_generate_object_from():
    _, _, database = rerun_data_from_files(["data/ats_start_port_ang_coos_bay.mqtt.json.bz2"],
                                           'test_save_job')
    model = Model()
    work_log = WorkLog(model, database)

    verify_work_log_as_json(work_log)


def test_save_2_freights():
    _, _, database = rerun_data_from_files(["data/ats_start_port_ang_coos_bay.mqtt.json.bz2",
                                            "data/ats_end_port_ang_coos_bay.mqtt.json.bz2",
                                            "data/ats_start_coos_dalles_short.mqtt.json.bz2",
                                            "data/ats_end_coos_dalles_short.mqtt.json.bz2"
                                            ], 'test_2_freights')
    lines = []
    for line in database._conn.iterdump():
        lines.append(line)
    verify('\n'.join(lines))


def test_load_2_freights():
    _, _, database = rerun_data_from_files(["data/ats_start_port_ang_coos_bay.mqtt.json.bz2",
                                            "data/ats_end_port_ang_coos_bay.mqtt.json.bz2",
                                            "data/ats_start_coos_dalles_short.mqtt.json.bz2",
                                            "data/ats_end_coos_dalles_short.mqtt.json.bz2"
                                            ], 'test_2_freights')
    model = Model()
    work_log = WorkLog(model, database)
    verify_work_log_as_json(work_log)