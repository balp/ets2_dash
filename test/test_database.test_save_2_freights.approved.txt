BEGIN TRANSACTION;
CREATE TABLE job
            (
                id      integer primary key,
                started integer,
                ended   integer
            );
INSERT INTO "job" VALUES(1,37201,37228);
INSERT INTO "job" VALUES(2,37228,38312);
INSERT INTO "job" VALUES(3,38312,38316);
INSERT INTO "job" VALUES(4,38316,38667);
INSERT INTO "job" VALUES(5,38667,NULL);
CREATE TABLE job_cancelled
            (
                id      integer primary key,
                penalty integer,
                FOREIGN KEY (id)
                    REFERENCES job (id)
            );
INSERT INTO "job_cancelled" VALUES(4,15000);
CREATE TABLE job_config
            (
                id                     integer primary key,
                cargo                  text,
                cargo_id               text,
                cargo_mass             real,
                delivery_time          integer,
                destination_city       text,
                destination_city_id    text,
                destination_company    text,
                destination_company_id text,
                income                 integer,
                source_city            text,
                source_city_id         text,
                source_company         text,
                source_company_id      text,
                FOREIGN KEY (id)
                    REFERENCES job (id)
            );
INSERT INTO "job_config" VALUES(2,'Lumber','lumber_b',21387.240234375,38821,'Coos Bay','coos_bay','Olthon Homes','oh_con_hom',18742,'Port Angeles','port_angeles','Deepgrove','dg_wd_saw');
INSERT INTO "job_config" VALUES(4,'Gypsum','gypsum',21492.798828125,38977,'The Dalles','the_dalles','Home Store','hs_mkt',12429,'Coos Bay','coos_bay','Home Store','hs_whs');
CREATE TABLE job_delivered
            (
                id           integer primary key,
                auto_load    integer,
                auto_park    integer,
                cargo_damage real,
                time         integer,
                distance     real,
                xp           integer,
                revenue      integer,
                FOREIGN KEY (id)
                    REFERENCES job (id)
            );
INSERT INTO "job_delivered" VALUES(2,1,0,0.0,1084,676.0,870,18742);
INSERT INTO "job_delivered" VALUES(4,1,0,0.0,351,487.0,795,12429);
CREATE TABLE track_point
            (
                id        integer,
                count     integer,
                time      integer,
                x real,
                y real,
                z real,
                PRIMARY KEY (id, count),
                FOREIGN KEY (id)
                    REFERENCES tracks (id)
            );
INSERT INTO "track_point" VALUES(1,0,37201,-1.04305182922363281247e+05,1.21028461456298828122e+01,-6.66552018508911132792e+04);
INSERT INTO "track_point" VALUES(1,1,37228,-1.04272021575927734376e+05,1.21054849624633789064e+01,-6.66048666954040527324e+04);
INSERT INTO "track_point" VALUES(2,0,37228,-1.04272021575927734376e+05,1.21054849624633789064e+01,-6.66048666954040527324e+04);
INSERT INTO "track_point" VALUES(2,1,38299,-1.13779683311462402345e+05,5.17771244049072265625e+00,-4.25592023086547851552e+04);
INSERT INTO "track_point" VALUES(2,2,38312,-1.13934719177246093746e+05,5.177143096923828125e+00,-4.27773633117675781229e+04);
INSERT INTO "track_point" VALUES(3,0,38312,-1.13934719177246093746e+05,5.17713594436645507812e+00,-4.27773633117675781229e+04);
INSERT INTO "track_point" VALUES(3,1,38315,-1.13746278945922851561e+05,5.17728519439697265625e+00,-4.28034440002441406245e+04);
INSERT INTO "track_point" VALUES(3,2,38316,-1.1373170094299316406e+05,5.18303823471069335937e+00,-4.27792263488769531265e+04);
INSERT INTO "track_point" VALUES(4,0,38316,-1.1373170094299316406e+05,5.18303823471069335937e+00,-4.27792263488769531265e+04);
INSERT INTO "track_point" VALUES(4,1,38318,-1.13712379978179931635e+05,5.17638063430786132812e+00,-4.27364988098144531246e+04);
INSERT INTO "track_point" VALUES(4,2,38659,-9.9166456390380859371e+04,1.22307882308959960939e+01,-5.1001428680419921874e+04);
INSERT INTO "track_point" VALUES(4,3,38660,-99233.9595336914,1.22502803802490234375e+01,-5.09348327941894531232e+04);
INSERT INTO "track_point" VALUES(4,4,38667,-99288.8861694336,1.23765115737915039058e+01,-5.10521015014648437508e+04);
INSERT INTO "track_point" VALUES(5,0,38667,-99288.8861694336,1.23764705657958984375e+01,-5.10521015014648437508e+04);
CREATE TABLE tracks
            (
                id    integer primary key,
                last_time integer,
                FOREIGN KEY (id)
                    REFERENCES job (id)
            );
INSERT INTO "tracks" VALUES(1,37228);
INSERT INTO "tracks" VALUES(2,38312);
INSERT INTO "tracks" VALUES(3,38316);
INSERT INTO "tracks" VALUES(4,38667);
INSERT INTO "tracks" VALUES(5,38667);
COMMIT;